### PR TITLE
HSEARCH-3692 Document custom bridges

### PR DESCRIPTION
**Pull request by yrodiere**
https://hibernate.atlassian.net/browse/HSEARCH-3692

I couldn't finish the documentation in time, but I think I got the most important parts covered.
I added follow-up tickets for the remainder.

Also, yes, I do realize DSL converters and projection converters need some love. Created [HSEARCH-3712](https://hibernate.atlassian.net/browse/HSEARCH-3712) to address this.

### CODE REVIEW

**Comment by yrodiere** [comment]
@gsmet Fabio offered to review now since he's available.

File: documentation/src/main/asciidoc/mapper-orm-bridge.asciidoc [hunk 1]
```diff
@@ -0,0 +1,1353 @@
...
+----
+<1> Map the property to the document identifier.
+<2> Instruct Hibernate Search to use our custom identifier bridge.
+It is also possible to reference the bridge by its name, in the case of a CDI/Spring bean.
+====
+
+=== Type resolution
+
+By default, the identifier bridge's property type is determined automatically,
+using reflection to extract the generic type argument of the `Identifier` interface.
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/mapper-orm-bridge.asciidoc]: Maybe `IdentifierBridge` interface?

File: documentation/src/main/asciidoc/mapper-orm-bridge.asciidoc [hunk 2]
```diff
@@ -0,0 +1,1353 @@
...
+using an <<mapper-orm-bridge-identifierbridge-identifierbinder,`IdentifierBinder`>>.
+
+=== Compatibility across indexes with `isCompatibleWith()`
+
+An identifier bridge is involved in indexing,
+but also in the search DSLs,
+to convert values passed to the <<search-dsl-predicate-id,`id` predicate>>
+to a document identifier that the backend will understand.
+
+When creating an `id` predicate targeting multiple indexes,
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/mapper-orm-bridge.asciidoc]: We're describing the mapping here. So maybe we can say 
s/targeting multiple indexes/targeting multiple entities?

File: documentation/src/main/asciidoc/mapper-orm-bridge.asciidoc [hunk 3]
```diff
@@ -0,0 +1,1353 @@
...
+
+Because of this complexity,
+when field types must be defined in the various binders
+(`ValueBinder`, `PropertyBinder`, `TypeBinder`),
+they are defined using a dedicated DSL.
+
+The entry point to this DSL is the `IndexFieldTypeFactory`.
+The type factory is generally accessible though the context object passed to the binders
+(`context.getTypeFactory()`).
+In the case of `PropertyBinder`, `TypeBinder`,
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/mapper-orm-bridge.asciidoc]: \,\ and \s?

File: documentation/src/main/asciidoc/mapper-orm-bridge.asciidoc [hunk 4]
```diff
@@ -0,0 +1,1353 @@
...
+
+.Defining a field type
+====
+[source, JAVA, indent=0, subs="+callouts"]
+----
+include::{sourcedir}/org/hibernate/search/documentation/mapper/orm/bridge/types/dsl/simple/ISBNBinder.java[tags=include]
+----
+<1> Get the `IndexFieldTypeFactory` from the binding context.
+<2> Define the data type.
+<3> Define options.
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/mapper-orm-bridge.asciidoc]: Optionally, we can add that available options depend on the data type. That we cannot have `normalizer` for a `double` type.

File: documentation/src/main/asciidoc/mapper-orm-bridge.asciidoc [hunk 5]
```diff
@@ -0,0 +1,1353 @@
...
+See <<mapper-orm-directfieldmapping-annotation-attributes>> for details about them.
+
+Other options are explained in the following sections.
+
+=== DSL converter
+
+[NOTE]
+====
+This section is not relevant for `ValueBinder`:
+Hibernate Search sets the DSL converter automatically for value bridges.
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/mapper-orm-bridge.asciidoc]: mabye we can add: taking their values form the annotations. ?

File: documentation/src/main/asciidoc/mapper-orm-bridge.asciidoc [hunk 6]
```diff
@@ -0,0 +1,1353 @@
...
+The various search DSLs expose some methods that expect a field value:
+`matching()`, `between()`, `atMost()`, `missingValue().use()`, ...
+By default, the expected type will be the same as the data type,
+i.e. `String` if you called `asString()`,
+`LocalDate` if you called `asLocalDate()`,
+etc.
+
+This can be annoying when the bridge converts values from a different type when indexing.
+For example, if the bridge converts an enum to a string when indexing,
+you probably don't want to pass string to search predicates,
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/mapper-orm-bridge.asciidoc]: a string?

**Comment by yrodiere** [comment]
Thanks! I addressed your comments and force-pushed. Waiting for CI.

**Pull request merged**